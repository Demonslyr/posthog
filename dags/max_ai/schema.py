from pydantic import BaseModel
from pydantic_avro import AvroBase

from posthog.schema import EventTaxonomyItem, TeamTaxonomyItem


# posthog/models/property_definition.py
class PropertyDefinitionSchema(AvroBase):
    name: str
    is_numerical: bool
    property_type: str | None
    type: int
    group_type_index: int | None


# posthog/models/warehouse/table.py
class DataWarehouseTableSchema(AvroBase):
    name: str
    format: str
    columns: list[str]


class PostgresProjectDataSnapshot(BaseModel):
    property_definitions: str
    data_warehouse_tables: str


# posthog/hogql_queries/ai/team_taxonomy_query_runner.py
class TeamTaxonomyItemSchema(AvroBase):
    results: list[TeamTaxonomyItem]


# posthog/hogql_queries/ai/event_taxonomy_query_runner.py
class PropertyTaxonomySchema(AvroBase):
    event: str
    results: list[EventTaxonomyItem]


class ClickhouseProjectDataSnapshot(BaseModel):
    event_taxonomy: str
    properties_taxonomy: str
