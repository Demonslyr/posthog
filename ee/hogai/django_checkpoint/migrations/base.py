"""
Simplified base class for checkpoint migrations.

Key simplification: Migrations only transform data, they don't touch the database.
This makes them simple, synchronous, and testable.
"""

from abc import ABC, abstractmethod
from typing import Any


class BaseMigration(ABC):
    """
    Base class for simple data migrations.

    Migrations should be deterministic, and only deal with data transformation, no database operations.
    Migrations are applied in the checkpoint serializer during deserialization.
    """

    @abstractmethod
    def migrate_data(self, data: dict[str, Any], type_hint: str) -> tuple[dict[str, Any], str]:
        """
        Apply the migration to a data dictionary.

        This is a pure function that transforms data. It should:
        - Not perform any I/O operations
        - Not access the database
        - Be idempotent (safe to run multiple times)
        - Handle missing fields gracefully
        - Can change the type hint for class renames

        Args:
            data: The data dictionary to migrate
            type_hint: The type hint (e.g., "AssistantState", "PartialAssistantState")
                      to help determine if migration applies

        Returns:
            Tuple of (migrated_data, new_type_hint)

        Example:
            # Rename a class
            if type_hint == "OldStateName":
                type_hint = "NewStateName"

            # Add a field
            if type_hint == "AssistantState" and "new_field" not in data:
                data["new_field"] = "default_value"

            return data, type_hint
        """
        pass
