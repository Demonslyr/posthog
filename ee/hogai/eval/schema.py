import json
from abc import ABC, abstractmethod
from collections.abc import Generator, Sequence
from typing import Generic, Self, TypeVar

from django.db.models import Model
from pydantic import BaseModel
from pydantic_avro import AvroBase

from posthog.models import DataWarehouseTable, PropertyDefinition, Team
from posthog.schema import EventTaxonomyItem, TeamTaxonomyItem


class EvalsDockerImageConfig(BaseModel):
    class Config:
        extra = "allow"

    bucket_name: str
    endpoint_url: str
    project_snapshots: list["Snapshot"]


T = TypeVar("T", bound=Model)


class BaseSchema(AvroBase, ABC, Generic[T]):
    @classmethod
    @abstractmethod
    def serialize_for_project(cls, project_id: int) -> Generator[Self, None, None]:
        raise NotImplementedError

    @classmethod
    @abstractmethod
    def deserialize_for_project(cls, project_id: int, models: Sequence[Self]) -> Generator[T, None, None]:
        raise NotImplementedError


class TeamSchema(BaseSchema[Team]):
    name: str
    test_account_filters: str

    @classmethod
    def serialize_for_project(cls, project_id: int):
        team = Team.objects.get(pk=project_id)
        yield TeamSchema(name=team.name, test_account_filters=json.dumps(team.test_account_filters))

    @classmethod
    def deserialize_for_project(cls, project_id: int, models: Sequence[Self]) -> Generator[Team, None, None]:
        for model in models:
            yield Team(id=project_id, name=model.name, test_account_filters=json.loads(model.test_account_filters))


# posthog/models/property_definition.py
class PropertyDefinitionSchema(BaseSchema[PropertyDefinition]):
    name: str
    is_numerical: bool
    property_type: str | None
    type: int
    group_type_index: int | None

    @classmethod
    def serialize_for_project(cls, project_id: int):
        for prop in PropertyDefinition.objects.filter(project_id=project_id).iterator(500):
            yield PropertyDefinitionSchema(
                name=prop.name,
                is_numerical=prop.is_numerical,
                property_type=prop.property_type,
                type=prop.type,
                group_type_index=prop.group_type_index,
            )

    @classmethod
    def deserialize_for_project(cls, project_id: int, models: Sequence[Self]):
        for model in models:
            yield PropertyDefinition(
                name=model.name,
                is_numerical=model.is_numerical,
                property_type=model.property_type,
                type=model.type,
                group_type_index=model.group_type_index,
                team_id=project_id,
            )


# posthog/models/warehouse/table.py
class DataWarehouseTableSchema(BaseSchema[DataWarehouseTable]):
    name: str
    format: str
    columns: list[str]

    @classmethod
    def serialize_for_project(cls, project_id: int):
        for table in DataWarehouseTable.objects.filter(team_id=project_id).iterator(500):
            yield DataWarehouseTableSchema(
                name=table.name,
                format=table.format,
                columns=table.columns,
            )

    @classmethod
    def deserialize_for_project(cls, project_id: int, models: Sequence[Self]):
        for model in models:
            yield DataWarehouseTable(
                name=model.name,
                format=model.format,
                columns=model.columns,
                url_pattern="http://localhost",
                team_id=project_id,
            )


class PostgresProjectDataSnapshot(BaseModel):
    project: str
    property_definitions: str
    data_warehouse_tables: str


# posthog/hogql_queries/ai/team_taxonomy_query_runner.py
class TeamTaxonomyItemSchema(AvroBase):
    results: list[TeamTaxonomyItem]


# posthog/hogql_queries/ai/event_taxonomy_query_runner.py
class PropertyTaxonomySchema(AvroBase):
    event: str
    results: list[EventTaxonomyItem]


class ClickhouseProjectDataSnapshot(BaseModel):
    event_taxonomy: str
    properties_taxonomy: str


class ProjectSnapshot(BaseModel):
    team_snapshot: TeamTaxonomyItemSchema
    postgres_snapshot: PostgresProjectDataSnapshot
    clickhouse_snapshot: ClickhouseProjectDataSnapshot


class Snapshot(BaseModel):
    project: int
    postgres: PostgresProjectDataSnapshot
    clickhouse: ClickhouseProjectDataSnapshot
