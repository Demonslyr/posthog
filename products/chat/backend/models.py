from typing import TYPE_CHECKING

from django.db import models
from django.db.models import QuerySet

from posthog.models.utils import UUIDModel
from posthog.models.file_system.file_system_mixin import FileSystemSyncMixin
from posthog.models.file_system.file_system_representation import FileSystemRepresentation

if TYPE_CHECKING:
    from posthog.models.team import Team


class ChatConversation(FileSystemSyncMixin, UUIDModel):
    class Meta:
        db_table = "posthog_chat_conversation"
        # No need for the unique constraint since we removed conversation_id
        pass

    person = models.ForeignKey(
        "posthog.Person",
        on_delete=models.CASCADE,
        related_name="chat_conversations",
        related_query_name="chat_conversation",
    )

    team = models.ForeignKey(
        "posthog.Team",
        on_delete=models.CASCADE,
        related_name="all_chat_conversations",
        related_query_name="all_chat_conversation",
    )

    title = models.CharField(max_length=400, blank=True, null=True)

    # Metadata
    created_at = models.DateTimeField(auto_now_add=True)
    updated_at = models.DateTimeField(auto_now=True)

    # Which page/url the conversation started on
    source_url = models.URLField(max_length=2000, blank=True, null=True)

    # Unread message count
    unread_count = models.PositiveIntegerField(default=0)

    @classmethod
    def get_file_system_unfiled(cls, team: "Team") -> QuerySet["ChatConversation"]:
        base_qs = cls.objects.filter(team=team)
        return cls._filter_unfiled_queryset(base_qs, team, type="chat_conversation", ref_field="id")

    def get_file_system_representation(self) -> FileSystemRepresentation:
        return FileSystemRepresentation(
            base_folder=self._create_in_folder or "Unfiled/Chats",
            type="chat_conversation",
            ref=str(self.pk),
            name=self.title
            or f"Chat with {self.person.properties.get('name') or self.person.properties.get('email') or f'Person {self.person.id}'}",
            href=f"/chat/{self.pk}",
            meta={
                "created_at": str(self.created_at),
                "person_id": str(self.person.id),
            },
            should_delete=False,
        )


class ChatMessage(UUIDModel):
    conversation = models.ForeignKey(
        ChatConversation,
        on_delete=models.CASCADE,
        related_name="messages",
        related_query_name="message",
    )

    content = models.TextField()

    # Metadata
    created_at = models.DateTimeField(auto_now_add=True)

    # Message has been read by the other party
    read = models.BooleanField(default=False)

    is_assistant = models.BooleanField(default=False, db_column="is_assistant")

    class Meta:
        ordering = ["created_at"]
        db_table = "posthog_chat_message"
