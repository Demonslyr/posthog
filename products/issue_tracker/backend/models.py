from django.db import models
from django.utils import timezone
import uuid


class Issue(models.Model):
    class Status(models.TextChoices):
        BACKLOG = "backlog", "Backlog"
        TODO = "todo", "To Do"
        IN_PROGRESS = "in_progress", "In Progress"
        TESTING = "testing", "Testing"
        DONE = "done", "Done"

    class OriginProduct(models.TextChoices):
        ERROR_TRACKING = "error_tracking", "Error Tracking"
        EVAL_CLUSTERS = "eval_clusters", "Eval Clusters"
        USER_CREATED = "user_created", "User Created"
        SUPPORT_QUEUE = "support_queue", "Support Queue"

    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
    team = models.ForeignKey("posthog.Team", on_delete=models.CASCADE)
    title = models.CharField(max_length=255)
    description = models.TextField()
    status = models.CharField(max_length=20, choices=Status.choices, default=Status.BACKLOG)
    origin_product = models.CharField(max_length=20, choices=OriginProduct.choices)
    position = models.IntegerField(default=0)
    created_at = models.DateTimeField(default=timezone.now)
    updated_at = models.DateTimeField(auto_now=True)

    class Meta:
        db_table = "posthog_issue"
        managed = True
        ordering = ["position"]

    def __str__(self):
        return f"{self.title} ({self.get_status_display()})"
