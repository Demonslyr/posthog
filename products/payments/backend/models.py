from typing import TYPE_CHECKING
from django.db import models
from posthog.models.utils import UUIDModel, sane_repr

if TYPE_CHECKING:
    pass


class PaymentProduct(UUIDModel):
    class Meta:
        db_table = "posthog_payments_products"

    team = models.ForeignKey("posthog.Team", on_delete=models.CASCADE)
    stripe_product_id = models.CharField(max_length=100)
    stripe_pricing_id = models.CharField(max_length=100)
    name = models.CharField(max_length=200)
    description = models.TextField(blank=True)
    price = models.DecimalField(max_digits=10, decimal_places=2)
    currency = models.CharField(max_length=3)  # e.g. USD, EUR
    date_created = models.DateTimeField(auto_now_add=True)
    date_updated = models.DateTimeField(auto_now=True)
    deleted = models.BooleanField(default=False)
    is_active = models.BooleanField(default=True)

    def __str__(self) -> str:
        return self.name

    __repr__ = sane_repr("id", "name", "team_id", "stripe_product_id")


class PaymentTransaction(UUIDModel):
    class Meta:
        db_table = "posthog_payments_transactions"

    team = models.ForeignKey("posthog.Team", on_delete=models.CASCADE)
    product = models.ForeignKey(PaymentProduct, on_delete=models.PROTECT)
    payload = models.JSONField()
    status = models.CharField(max_length=50)
    date_created = models.DateTimeField(auto_now_add=True)

    def __str__(self) -> str:
        return f"{self.team_id} - {self.product.name} - {self.status}"

    __repr__ = sane_repr("id", "team_id", "product_id", "status")
