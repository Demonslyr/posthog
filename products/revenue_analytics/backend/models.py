from typing import cast, Any, Optional
import posthoganalytics

from posthog.hogql import ast
from posthog.warehouse.models.external_data_source import ExternalDataSource
from posthog.warehouse.models.external_data_schema import ExternalDataSchema
from posthog.warehouse.models.table import DataWarehouseTable
from posthog.models.exchange_rate.sql import EXCHANGE_RATE_DECIMAL_PRECISION
from posthog.schema import CurrencyCode
from posthog.hogql.database.models import (
    BooleanDatabaseField,
    DateTimeDatabaseField,
    ExpressionField,
    IntegerDatabaseField,
    StringDatabaseField,
    Table,
    FieldOrTable,
)
from posthog.hogql.database.schema.exchange_rate import (
    DEFAULT_CURRENCY,
    convert_currency_call,
)


from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from posthog.hogql.context import HogQLContext

STRIPE_DATA_WAREHOUSE_CHARGE_IDENTIFIER = "Charge"

# Stripe represents most currencies with integer amounts multiplied by 100,
# since most currencies have its smallest unit as 1/100 of their base unit
# It just so happens that some currencies don't have that concept, so they're listed here
# https://docs.stripe.com/currencies#zero-decimal
ZERO_DECIMAL_CURRENCIES_IN_STRIPE: list[CurrencyCode] = [
    CurrencyCode.BIF,
    CurrencyCode.CLP,
    CurrencyCode.DJF,
    CurrencyCode.GNF,
    CurrencyCode.JPY,
    CurrencyCode.KMF,
    CurrencyCode.KRW,
    CurrencyCode.MGA,
    CurrencyCode.PYG,
    CurrencyCode.RWF,
    CurrencyCode.UGX,
    CurrencyCode.VND,
    CurrencyCode.VUV,
    CurrencyCode.XAF,
    CurrencyCode.XOF,
    CurrencyCode.XPF,
]

BASE_FIELDS: dict[str, FieldOrTable] = {
    "distinct_id": StringDatabaseField(name="id"),
    "__original_amount": IntegerDatabaseField(name="amount", hidden=True),
    "original_amount": ExpressionField(
        isolate_scope=True,
        expr=ast.Call(
            name="toDecimal",
            args=[
                ast.Field(chain=["__original_amount"]),
                ast.Constant(value=EXCHANGE_RATE_DECIMAL_PRECISION),
            ],
        ),
        name="original_amount",
    ),
    "is_refund": BooleanDatabaseField(name="refunded"),
    "__original_currency": StringDatabaseField(name="currency", hidden=True),
    "original_currency": ExpressionField(
        isolate_scope=True,
        expr=ast.Call(name="upper", args=[ast.Field(chain=["__original_currency"])]),
        name="original_currency",
    ),
    # As per above, Stripe represents most currencies with integer amounts multiplied by 100,
    # so we need to divide by 100 to get the actual amount
    "currency_is_zero_decimal": ExpressionField(
        isolate_scope=True,
        expr=ast.CompareOperation(
            op=ast.CompareOperationOp.In,
            left=ast.Field(chain=["original_currency"]),
            right=ast.Array(
                exprs=[ast.Constant(value=currency.value) for currency in ZERO_DECIMAL_CURRENCIES_IN_STRIPE]
            ),
        ),
        name="currency_is_zero_decimal",
        hidden=True,
    ),
    "amount_decimal_divider": ExpressionField(
        isolate_scope=True,
        expr=ast.Call(
            name="if",
            args=[
                ast.Field(chain=["currency_is_zero_decimal"]),
                ast.Call(
                    name="toDecimal", args=[ast.Constant(value=1), ast.Constant(value=EXCHANGE_RATE_DECIMAL_PRECISION)]
                ),
                ast.Call(
                    name="toDecimal",
                    args=[ast.Constant(value=100), ast.Constant(value=EXCHANGE_RATE_DECIMAL_PRECISION)],
                ),
            ],
        ),
        name="amount_decimal_divider",
        hidden=True,
    ),
    # Amount taking into consideration whether this was a refund or not
    # plus the decimal divider
    "adjusted_original_amount": ExpressionField(
        isolate_scope=True,
        expr=ast.Call(
            name="divideDecimal",
            args=[
                ast.Call(
                    name="if",
                    args=[
                        ast.Field(chain=["is_refund"]),
                        ast.Call(name="negate", args=[ast.Field(chain=["original_amount"])]),
                        ast.Field(chain=["original_amount"]),
                    ],
                ),
                ast.Field(chain=["amount_decimal_divider"]),
            ],
        ),
        name="adjusted_original_amount",
    ),
    # Allow us to sort by timestamp properly in insights (and split in buckets)
    "timestamp": DateTimeDatabaseField(name="created_at"),
}


class RevenueAnalyticsRevenueView(Table):
    data_warehouse_table: (
        Any  # Pydantic is complaining for reasons I dont understand, this should be DataWarehouseTable
    )

    @staticmethod
    def for_schema_source(source: ExternalDataSource) -> Optional["RevenueAnalyticsRevenueView"]:
        # Currently only works for stripe sources
        if not source.source_type == ExternalDataSource.Type.STRIPE:
            return None

        # The table we care about is the one with schema `Charge` since from there we can get
        # the data we need in our view
        try:
            schema: ExternalDataSchema = ExternalDataSchema.objects.get(
                source=source, name=STRIPE_DATA_WAREHOUSE_CHARGE_IDENTIFIER
            )
            table: Optional[DataWarehouseTable] = cast(
                Optional[DataWarehouseTable], schema.table
            )  # Weird cast because pydantic is weird
        except (ExternalDataSchema.DoesNotExist, DataWarehouseTable.DoesNotExist):
            return None

        if table is None:
            return None

        team = table.team
        revenue_config = team.revenue_config
        do_currency_conversion = posthoganalytics.feature_enabled(
            "web-analytics-revenue-tracking-conversion",
            str(team.organization_id),
            groups={"organization": str(team.organization_id)},
            group_properties={"organization": {"id": str(team.organization_id)}},
        )

        # Get most of the fields directly from the base fields,
        # but generate the converted currency and amounts dynamically
        # since we depend on the account's configuration
        fields: dict[str, FieldOrTable] = {
            **BASE_FIELDS,
            "currency": ExpressionField(
                isolate_scope=True,
                expr=ast.Constant(value=(revenue_config.baseCurrency or DEFAULT_CURRENCY).value),
                name="currency",
            )
            if do_currency_conversion
            else BASE_FIELDS["original_currency"],
            "amount": ExpressionField(
                isolate_scope=True,
                expr=convert_currency_call(
                    # Only doing the adjusted calculation here since this is after we've divided by 100
                    ast.Field(chain=["adjusted_original_amount"]),
                    ast.Field(chain=["original_currency"]),
                    ast.Field(chain=["currency"]),
                    ast.Call(
                        name="_toDate",
                        args=[
                            # Because timestamp columns are nullable, we need to handle that case
                            # by converting to a default value of 0
                            ast.Call(
                                name="ifNull",
                                args=[
                                    ast.Field(chain=["timestamp"]),
                                    ast.Call(name="toDateTime", args=[ast.Constant(value=0)]),
                                ],
                            )
                        ],
                    ),
                ),
                name="amount",
            )
            if do_currency_conversion
            else BASE_FIELDS["adjusted_original_amount"],
            # Required because we add a "team_id" where clause on every selected table
            "team_id": ExpressionField(
                isolate_scope=True,
                expr=ast.Constant(value=team.pk),
                name="team_id",
            ),
        }

        return RevenueAnalyticsRevenueView(fields=fields, data_warehouse_table=table)

    def to_printed_clickhouse(self, context: "HogQLContext") -> str:
        table = cast(DataWarehouseTable, self.data_warehouse_table)
        return table.name

    def to_printed_hogql(self) -> str:
        table = cast(DataWarehouseTable, self.data_warehouse_table)
        return table.name
