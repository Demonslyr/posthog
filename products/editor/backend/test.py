import json
from datetime import datetime

from pydantic import BaseModel, Field

from posthog.clickhouse.client.execute import sync_execute
from posthog.test.base import BaseTest
from products.editor.backend.models.codebase import Codebase
from products.editor.backend.services.codebase_sync import SerializedArtifact


class CatalogEntry(BaseModel):
    team_id: int | None = Field(default=None)
    user_id: int | None = Field(default=None)
    codebase_id: str | None = Field(default=None)
    artifact_id: str
    parent_artifact_id: str | None = Field(default=None)
    branch: str | None = Field(default=None)
    type: str = Field(default="dir")
    sign: int = Field(default=1)
    timestamp: datetime | None = Field(default=None)


class EditorTestQueryHelpersMixin(BaseTest):
    codebase: Codebase
    stable_user_id: int = 99999

    def _create_artifacts(self, tree: list[SerializedArtifact]):
        query = "INSERT INTO codebase_embeddings (team_id, user_id, codebase_id, artifact_id, chunk_id, vector, properties, is_deleted) VALUES "
        rows: list[str] = []

        args = {
            "team_id": self.team.id,
            "user_id": self.user.id,
            "codebase_id": self.codebase.id,
            "chunk_id": 0,
            "vector": [0.5, 0.5],
            "properties": json.dumps(
                {
                    "lineStart": 0,
                    "lineEnd": 30,
                    "path": "obfuscated_path",
                }
            ),
            "is_deleted": 0,
        }

        for idx, node in enumerate(tree):
            if node["type"] == "file":
                args.update(
                    {
                        f"artifact_id_{idx}": node["id"],
                    }
                )
                rows.append(
                    f"(%(team_id)s, %(user_id)s, %(codebase_id)s, %(artifact_id_{idx})s, %(chunk_id)s, %(vector)s, %(properties)s, %(is_deleted)s)"
                )

        sync_execute(query + ", ".join(rows), args, team_id=self.team.id)

    def _create_codebase_catalog(self, tree: list[CatalogEntry]):
        query = "INSERT INTO codebase_catalog (team_id, user_id, codebase_id, artifact_id, parent_artifact_id, branch, type, timestamp, sign) VALUES "
        rows: list[str] = []
        args = {}

        for idx, node in enumerate(tree):
            args.update(
                {
                    f"team_id_{idx}": node.team_id or self.team.id,
                    f"user_id_{idx}": node.user_id or self.stable_user_id,
                    f"codebase_id_{idx}": node.codebase_id or str(self.codebase.id),
                    f"artifact_id_{idx}": node.artifact_id,
                    f"parent_artifact_id_{idx}": node.parent_artifact_id,
                    f"branch_{idx}": node.branch or "",
                    f"type_{idx}": node.type,
                    f"timestamp_{idx}": node.timestamp,
                    f"sign_{idx}": node.sign,
                }
            )
            rows.append(
                f"(%(team_id_{idx})s, %(user_id_{idx})s, %(codebase_id_{idx})s, %(artifact_id_{idx})s, %(parent_artifact_id_{idx})s, %(branch_{idx})s, %(type_{idx})s, %(timestamp_{idx})s, %(sign_{idx})s)"
            )

        sync_execute(query + ", ".join(rows), args, team_id=self.team.id)
