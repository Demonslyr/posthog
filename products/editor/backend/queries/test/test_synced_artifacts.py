from datetime import datetime

from posthog.schema import SyncedArtifactsQuery, SyncedArtifactsResponseItem
from posthog.test.base import BaseTest, ClickhouseTestMixin, snapshot_clickhouse_queries
from products.editor.backend.models.codebase import Codebase
from products.editor.backend.test.base import EditorTestQueryHelpersMixin

from ..synced_artifacts import SyncedArtifactsQueryRunner


class TestSyncedArtifacts(EditorTestQueryHelpersMixin, ClickhouseTestMixin, BaseTest):
    def setUp(self):
        super().setUp()
        self.codebase = Codebase.objects.create(
            team=self.team, user=self.user, id="019634cb-5198-0000-ab58-7ba5d8855147"
        )

    @snapshot_clickhouse_queries
    def test_synced_artifacts(self):
        """Must return a single artifact."""
        self._create_artifacts([{"id": "file1", "type": "file", "parent_id": "root"}], user_id=self.stable_user_id)

        response = SyncedArtifactsQueryRunner(
            SyncedArtifactsQuery(userId=self.stable_user_id, codebaseId=str(self.codebase.id), artifactIds=["file1"]),
            self.team,
        ).run()
        self.assertCountEqual(response.results, [SyncedArtifactsResponseItem(id="file1")])

    @snapshot_clickhouse_queries
    def test_artifacts_from_the_list_are_returned(self):
        """Artifacts can be filtered by a list of artifact IDs."""
        self._create_artifacts(
            [
                {"id": "file1", "type": "file", "parent_id": "root"},
                {"id": "file2", "type": "file", "parent_id": "root"},
            ],
            user_id=self.stable_user_id,
        )

        response = SyncedArtifactsQueryRunner(
            SyncedArtifactsQuery(userId=self.stable_user_id, codebaseId=str(self.codebase.id), artifactIds=["file2"]),
            self.team,
        ).run()
        self.assertCountEqual(response.results, [SyncedArtifactsResponseItem(id="file2")])

    def test_handles_multiple_users(self):
        """Artifacts must be scoped to a user."""
        self._create_artifacts(
            [
                {"id": "file1", "type": "file", "parent_id": "root"},
            ],
            user_id=self.stable_user_id,
        )
        self._create_artifacts(
            [
                {"id": "file2", "type": "file", "parent_id": "root"},
            ],
            user_id=100000,
        )

        response = SyncedArtifactsQueryRunner(
            SyncedArtifactsQuery(
                userId=self.stable_user_id, codebaseId=str(self.codebase.id), artifactIds=["file1", "file2"]
            ),
            self.team,
        ).run()
        self.assertCountEqual(response.results, [SyncedArtifactsResponseItem(id="file1")])

        response = SyncedArtifactsQueryRunner(
            SyncedArtifactsQuery(userId=100000, codebaseId=str(self.codebase.id), artifactIds=["file1", "file2"]),
            self.team,
        ).run()
        self.assertCountEqual(response.results, [SyncedArtifactsResponseItem(id="file2")])

    def test_handles_multiple_codebases(self):
        """Artifacts can be filtered by a list of artifact IDs."""
        self._create_artifacts(
            [
                {"id": "file1", "type": "file", "parent_id": "root"},
                {"id": "file2", "type": "file", "parent_id": "root"},
            ],
            user_id=self.stable_user_id,
        )

        another_codebase = "019634cb-5198-0000-ab58-7ba5d8855148"
        self._create_artifacts(
            [
                {"id": "file1", "type": "file", "parent_id": "root"},
            ],
            user_id=self.stable_user_id,
            codebase_id=another_codebase,
        )

        response = SyncedArtifactsQueryRunner(
            SyncedArtifactsQuery(
                userId=self.stable_user_id, codebaseId=str(self.codebase.id), artifactIds=["file1", "file2"]
            ),
            self.team,
        ).run()
        self.assertCountEqual(
            response.results, [SyncedArtifactsResponseItem(id="file1"), SyncedArtifactsResponseItem(id="file2")]
        )

        response = SyncedArtifactsQueryRunner(
            SyncedArtifactsQuery(
                userId=self.stable_user_id, codebaseId=another_codebase, artifactIds=["file1", "file2"]
            ),
            self.team,
        ).run()
        self.assertCountEqual(response.results, [SyncedArtifactsResponseItem(id="file1")])

    def test_returns_distinct_artifacts(self):
        """Must return a single artifact across multiple chunks."""
        self._create_artifacts(
            [
                {"id": "file1", "type": "file", "parent_id": "root", "chunk_id": 0},
                {"id": "file1", "type": "file", "parent_id": "root", "chunk_id": 1},
            ],
            user_id=self.stable_user_id,
        )

        response = SyncedArtifactsQueryRunner(
            SyncedArtifactsQuery(userId=self.stable_user_id, codebaseId=str(self.codebase.id), artifactIds=["file1"]),
            self.team,
        ).run()
        self.assertCountEqual(response.results, [SyncedArtifactsResponseItem(id="file1")])

    def test_omits_deleted_artifacts(self):
        """Must omit deleted artifacts."""
        ts = datetime.now()
        self._create_artifacts(
            [
                {"id": "file1", "type": "file", "parent_id": "root", "is_deleted": 1, "timestamp": ts},
            ],
            user_id=self.stable_user_id,
        )

        response = SyncedArtifactsQueryRunner(
            SyncedArtifactsQuery(userId=self.stable_user_id, codebaseId=str(self.codebase.id), artifactIds=["file1"]),
            self.team,
        ).run()
        self.assertCountEqual(response.results, [])

    def test_omits_deleted_rows_based_on_the_last_timestamp(self):
        """Must return the latest artifact."""
        ts = datetime.now()
        self._create_artifacts(
            [
                {"id": "file1", "type": "file", "parent_id": "root", "is_deleted": 0, "timestamp": ts},
                {"id": "file1", "type": "file", "parent_id": "root", "is_deleted": 1, "timestamp": ts},
            ],
            user_id=self.stable_user_id,
        )

        response = SyncedArtifactsQueryRunner(
            SyncedArtifactsQuery(userId=self.stable_user_id, codebaseId=str(self.codebase.id), artifactIds=["file1"]),
            self.team,
        ).run()
        self.assertCountEqual(response.results, [])
