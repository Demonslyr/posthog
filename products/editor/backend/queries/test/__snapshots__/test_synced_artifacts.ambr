# serializer version: 1
# name: TestSyncedArtifacts.test_artifacts_from_the_list_are_returned
  '''
  select *
  from codebase_embeddings
  '''
# ---
# name: TestSyncedArtifacts.test_artifacts_from_the_list_are_returned.1
  '''
  SELECT codebase_embeddings.artifact_id AS synced_artifact_id,
         argMax(codebase_embeddings.is_deleted, toTimeZone(codebase_embeddings.timestamp, 'UTC')) AS deleted
  FROM codebase_embeddings
  WHERE and(equals(codebase_embeddings.team_id, 99999), equals(codebase_embeddings.user_id, 99999), equals(codebase_embeddings.codebase_id, '019634cb-5198-0000-ab58-7ba5d8855147'), in(codebase_embeddings.artifact_id, ['file2']))
  GROUP BY codebase_embeddings.artifact_id
  HAVING ifNull(equals(deleted, 0), 0)
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
# name: TestSyncedArtifacts.test_synced_artifacts
  '''
  select *
  from codebase_embeddings
  '''
# ---
# name: TestSyncedArtifacts.test_synced_artifacts.1
  '''
  SELECT codebase_embeddings.artifact_id AS synced_artifact_id,
         argMax(codebase_embeddings.is_deleted, toTimeZone(codebase_embeddings.timestamp, 'UTC')) AS deleted
  FROM codebase_embeddings
  WHERE and(equals(codebase_embeddings.team_id, 99999), equals(codebase_embeddings.user_id, 99999), equals(codebase_embeddings.codebase_id, '019634cb-5198-0000-ab58-7ba5d8855147'), in(codebase_embeddings.artifact_id, ['file1']))
  GROUP BY codebase_embeddings.artifact_id
  HAVING ifNull(equals(deleted, 0), 0)
  LIMIT 100 SETTINGS readonly=2,
                     max_execution_time=60,
                     allow_experimental_object_type=1,
                     format_csv_allow_double_quotes=0,
                     max_ast_elements=4000000,
                     max_expanded_ast_elements=4000000,
                     max_bytes_before_external_group_by=0,
                     transform_null_in=1,
                     optimize_min_equality_disjunction_chain_length=4294967295
  '''
# ---
